### ISSUE TITLE
Unable to create BC OnPrem container

### ISSUE THREAD

**Issue by Bleika** [Microsoft/navcontainerhelper#260]
I'm struggling to create my first NavContainer (for BC OnPrem ES), because the New-NavContainer always stalls (last successful output: Assign Premium Plan...), before throwing an error, trying to initialize again and finally aborting execution.

If I try to debug, I'm always taken to Wait-NavContainerReady where it starts outputting log content (basically saying "Warning: operation successful but took too long" on SQL issues). 
I don't get this output if I don't enter in debug mode (see below).

I'm using an up-to-date Windows Server 2016 on a Virtual machine.
I've already run troubleshooting scripts on my Docker EE installation to discard any installation-related issues, and they all come out fine.
This happens both in 0.4.0.2 and 0.4.1.2.

**I run the command with these parameters:**
PS C:\> New-NavContainer -containerName BCOnPremES -accept_eula -assignPremiumPlan -auth Windows -authenticationEMail X@X.es -Credential administrador -doNotExportObjectsToText -enableSymbolLoading -imageName mcr.microsoft.com/businesscentral/onprem:es -includeCSide -includeTestLibrariesOnly -includeTestToolkit -isolation process -shortcuts Desktop -timeout 7200 -updateHosts
(I increased timeout to 7200 to see if it would write the log Warnings to the console, and I don't have any personalized scripts in the shared folders).

**This is the output I get:**
Host is Windows Server 10.0.14393.0 - ltsc2016
Using image mcr.microsoft.com/businesscentral/onprem:es
Removing container BCOnPremES
Removing BCOnPremES from hosts
Removing C:\ProgramData\NavContainerHelper\Extensions\BCOnPremES
Creating Nav container BCOnPremES
NAV Version: 13.0.24630.0-es
Generic Tag: 0.0.6.6
Container OS Version: 10.0.14393.2430 (ltsc2016)
Host OS Version: 10.0.14393.0 (ltsc2016)
Creating container BCOnPremES from image mcr.microsoft.com/businesscentral/onprem:es
1491429e70a39f8b9e39309de07e9cf62bc314086939b34873e3694a5ba5484c
Waiting for container BCOnPremES to be ready
Initializing...
Starting Container
Hostname is BCOnPremES
PublicDnsName is BCOnPremES
Using Windows Authentication
Starting Local SQL Server
Starting Internet Information Server
Modifying Service Tier Config File with Instance Specific Settings
Starting NAV Service Tier
Creating DotNetCore Web Server Instance
Creating http download site
Creating Windows user administrador
Setting SA Password and enabling SA
Creating SUPER user
Assign Premium plan for BCONPREMES\ADMINISTRADOR
Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 172
at <ScriptBlock>, C:\Run\start.ps1: line 107
at <ScriptBlock>, <No file>: line 1Error
Initializing...
Starting Container
Hostname is BCOnPremES
PublicDnsName is BCOnPremES
Using Windows Authentication
Starting Local SQL Server
Starting Internet Information Server
Modifying Service Tier Config File with Instance Specific Settings
Starting NAV Service Tier
Creating DotNetCore Web Server Instance
Creating http download site
Creating Windows user administrador
Setting SA Password and enabling SA
Creating SUPER user
Assign Premium plan for BCONPREMES\ADMINISTRADOR
Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 172
at <ScriptBlock>, C:\Run\start.ps1: line 107
at <ScriptBlock>, <No file>: line 1
Initialization of container BCOnPremES failed
En C:\Program 
Files\WindowsPowerShell\Modules\navcontainerhelper\0.4.1.2\ContainerHandling\Wait-NavContainerReady.ps1: 
45 Carácter: 17
+ ...             throw "Initialization of container $containerName failed" ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Initialization ...OnPremES failed:String) [], RuntimeExc 
   eption
    + FullyQualifiedErrorId : Initialization of container BCOnPremES failed
 
Please advise

**Comment by NAVspecialist** [comment]
You are using a tons of parameters :) Have you tried a more simple approach like this ?

$Licensefile = "C:\Docker\YourLincense365.flf"
$ImageName = "mcr.microsoft.com/businesscentral/onprem:es"
$ContainerName = "nav365bc"
$PlainTextPw = "Passw0rd"
$Credential = New-Object -TypeName PSCredential -ArgumentList $env:USERNAME,(ConvertTo-SecureString -String $PlainTextPw -AsPlainText -Force)
New-NavContainer -containerName $ContainerName -memoryLimit "3g" -accept_eula -accept_outdated -auth NavUserPassword -Credential $Credential -imageName $ImageName -licenseFile $Licensefile -includeCSide -shortcuts Desktop -verbose

Palle

**Comment by freddydk** [comment]
Palle is right, your problem is probably some of the parameters - I guess it is -credential, it needs to be a pscredential - or leave it out, then new-navcontainer will ask you

**Comment by Bleika** [comment]
Tusind tak til begge to!

I had a look at the logs and found out they still contained entries from my former intents (this struggle has gone on for days now). Therefore I decided to have a clean start.
I've also pruned both containers and images before continuing and restarted the docker service.

**Following your instructions, I've done several attempts with simplified parameters, and I now think the parameters can be outruled as a cause to my problems. Now it stalls after "Creating SUPER user":**
(line breaks come out a bit strange on here, but they're right in PS).
PS C:\> $MyImageName = "mcr.microsoft.com/businesscentral/onprem:es"
$MyContainerName = "BCOnPremES4"

New-NavContainer -accept_eula `
                 -containerName $MyContainerName `
                 -imageName $MyImageName `
                 -auth Windows `
                 -enableSymbolLoading `
                 -includeCSide `
                 -doNotExportObjectsToText `
                 -shortcuts Desktop `
                 -updateHosts                
Host is Windows Server 10.0.14393.0 - ltsc2016
Using image mcr.microsoft.com/businesscentral/onprem:es
Creating Nav container BCOnPremES4
NAV Version: 13.0.24630.0-es
Generic Tag: 0.0.6.6
Container OS Version: 10.0.14393.2430 (ltsc2016)
Host OS Version: 10.0.14393.0 (ltsc2016)
Creating container BCOnPremES4 from image mcr.microsoft.com/businesscentral/onprem:es
5913853f32443c87f015efa876e5b9fc5c0aae274c8ba388f115435a933898ae
Waiting for container BCOnPremES4 to be ready
Initializing...
Starting Container
Hostname is BCOnPremES4
PublicDnsName is BCOnPremES4
Using Windows Authentication
Starting Local SQL Server
Starting Internet Information Server
Modifying Service Tier Config File with Instance Specific Settings
Starting NAV Service Tier
Creating DotNetCore Web Server Instance
Creating http download site
Creating Windows user administrador
Setting SA Password and enabling SA
Creating SUPER user
**Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 172
at <ScriptBlock>, C:\Run\start.ps1: line 107
at <ScriptBlock>, <No file>: line 1Error**
Initializing...
Starting Container
Hostname is BCOnPremES4
PublicDnsName is BCOnPremES4
Using Windows Authentication
Starting Local SQL Server
Starting Internet Information Server
Modifying Service Tier Config File with Instance Specific Settings
Starting NAV Service Tier
Creating DotNetCore Web Server Instance
Creating http download site
Creating Windows user administrador
Setting SA Password and enabling SA
Creating SUPER user
**Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 172
at <ScriptBlock>, C:\Run\start.ps1: line 107
at <ScriptBlock>, <No file>: line 1
Initialization of container BCOnPremES4 failed
En C:\Program Files\WindowsPowerShell\Modules\navcontainerhelper\0.4.1.2\ContainerHandling\Wait-NavContainerReady.ps1: 45 
Carácter: 17
+ ...             throw "Initialization of container $containerName failed" ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Initialization ...nPremES4 failed:String) [], RuntimeException
    + FullyQualifiedErrorId : Initialization of container BCOnPremES4 failed**
  
**Logs:**
I'm afraid my Host & Image are ES so most of it's in spanish.
I'm now contemplating creating a new VM in EN to start all over, but due to resource problems that probably won't be until tomorrow (I've been struggling like crazy to get some hands-on experience with this environment before NTD, but I suppose I must accept that I've run out of time).

- I've attached the "docker logs BCOnPremES4" output in .txt. It only contains 9 warnings and they all seem to originate from the same source: "Long running SQL statements".
**Could this be the cause of the array error?** 
[DockerLogsBCOnPremES4.txt](https://github.com/Microsoft/navcontainerhelper/files/2600379/DockerLogsBCOnPremES4.txt)

- I've also attached a more interesting .xlsx with the Event Log content for this last attempt (Get-NAVContainerEventLog BCOnPremES4). (Full: includes all entries. ErrorWarning: filtered).
The .xlsx also contains the host Event Log specifically for Docker in the same way.
[Event logs for BCOnPremES4.xlsx](https://github.com/Microsoft/navcontainerhelper/files/2600380/Event.logs.for.BCOnPremES4.xlsx)

As to container specific entries (.xlsx), they identify a few issues:
**- The   service could not add service principal names because the service account could not be found in Active Directory. Account NT AUTHORITY\SYSTEM (when defining ports 7045..7048.**
This surprises me a bit: in the spanish localization of Win, this account translates into "Sistema local". 
Is it possible that there could be some kind of a language mismatch when searching for the account?
(It definitely exists on my host VM and I use it for many running services).
- Several Warnings on "long running" SQL statement 
- Begin IsolationLevel=65536 is attempted several times
- License activation error (slui.exe)
- Automatic startup of DefAppPool failed

**Apart from the log findings, I doubt about a few things me and maybe either of you can help me clarify them:**
1) Docker installation didn't go as expected: I cannot start the service if I let Docker create my nat network automatically, so I had to include a { "bridge" : "none" } instruction in daemon.json and create my own nat manually. Furthermore, it only works if I specify subnet and gateway. (It's all within the hosts range and at least the troubleshooting script for docker accepts it and comes away 100% right).
I've gone over the different pre-requisites a dozen times and I don't understand why I need the "bridge" :"none" statement (maybe server language causes problems?).
Apart from harrassing FK on Yammer the other day (!!), I've also sought help on MSDN, but I'm still using "my personalized nat" network. (Inspect/NetNat details further below).

2) A few days ago, I successfully created a container, although based on the wrong image (sandbox). 
That time, the script was completed (I didn't get the errors I'm getting now), but a) I couldn't connect it to my network (inspite of using --attachable param) and b) it wasn't the image I wanted.
**Therefore, somehow I relate the current errors to the difference between sandbox and onprem, although I haven't been able to figure out why they happen (image?/deployment platform?/language?).**

3) The VM I'm using, already has NAV2018 installed (the traditional way) - I don't know if that can cause any serious conflict that might be preventing the container creation (as a Docker rookie I assumed I might perhaps face some port conflicts, etc. but would still be able to create the container and resolve any issues afterwards). At least it didn't prevent me from creating the sandbox container.
It also has a full SQL Server 2016 Standard installed (and not SQL Server Express).

**Docker network inspect output:**
        "Name": "bclan01",
        "Id": "cfedbd619131d23d78d20b872b747f46850e164e06d5f56000dcbc9c29bca1b7",
        "Created": "2018-11-20T09:58:28.7694311+01:00",
        "Scope": "local",
        "Driver": "nat",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "windows",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.27.16.0/24",
                    "Gateway": "172.27.16.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": true,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {},
        "Options": {
            "com.docker.network.windowsshim.hnsid": "271205af-bcfd-4e71-8f58-19a88efaba29",
            "com.docker.network.windowsshim.networkname": "bclan01"
        },
        "Labels": {}

**Get-NetNat:**
Name                             : Hf1e9e4b7-43e9-4954-b27d-da8d41e5987e
ExternalIPInterfaceAddressPrefix : 
InternalIPInterfaceAddressPrefix : 172.27.16.0/24
IcmpQueryTimeout                 : 30
TcpEstablishedConnectionTimeout  : 1800
TcpTransientConnectionTimeout    : 120
TcpFilteringBehavior             : AddressDependentFiltering
UdpFilteringBehavior             : AddressDependentFiltering
UdpIdleSessionTimeout            : 120
UdpInboundRefresh                : False
Store                            : Local
Active                           : True

Tak igen (og god rejse - jeg tror vi træffes på torsdag),
Eyrún

**Comment by freddydk** [comment]
If you try again and add -alwaysPull to your new-navcontianer - it will download cu1 and that will probably work.
The reason for the error is probably network setting, like this: https://github.com/Microsoft/nav-docker/issues/281
which was fixed in a later generic version - which was used for business central on prem cu1
All older versions will be updated with this fix as well.

**Comment by Bleika** [comment]
Tak igen Freddy og undskyld: jeg burde have fundet det selv...

The error on issue #281 is a close fit to my situation - the only difference is that my network problems are nat related.

I tried again with the following command:
$MyContainerName = "BCOnPremES5"
$MyDBServer = "BCServer"

New-NavContainer -accept_eula `
                 -containerName $MyContainerName `
                 -imageName $MyImageName `
                 -auth Windows `
                 -databaseServer $MyDBServer `
                 -enableSymbolLoading `
                 -includeCSide `
                 -doNotExportObjectsToText `
                 -shortcuts Desktop `
                 -timeout 7200 `
                 -updateHosts `
                 -alwaysPull `
                 -verbose

That went much better: I've gotten rid of the array error and now I only have some network related issues. 

**Most relative script output:**
WARNING: DNS resolution not working from within the container.
Container IP Address: UNKNOWN

I've run out of time to dig deeper into the networking issues - I'll continue after NTD.
Also, it won't let me include the TestToolkit related parameters (I suspect a mismatch in the server name), but it's not important right now. 

Thanks again,
Eyrún

**Comment by Bleika** [comment]
You're right - the -databaseServer parameter is superfluos in my case - I'll omit it in my next test (hopefully it'll avoid the TestToolkit error). 

The network issue in Docker is really annoying. 
I'd have thought that an up-to-date WinServer should be perfectly compatible and the install should have been a straight forward and easy task. More so, since I understand I'm deploying the simplest scenario.
However, even though all troubleshooting/inspect tests come out clean, I'm unable to get docker running unless I disable the automatic network creation (bridge:none).

Last news from MSDN is a suggestion to downgrade Docker to ver. 17.06 (I'm using 18.09).
I won't do that until I've had a chance to see if I can connect my new container to the personalized nat network.

**Comment by NAVspecialist** [comment]
Hi there

A few suggestions on the DNS-matter. What I have done is the following.

1) I ran SCONFIG as administrator and changed the DNS to 8.8.8.8. and 8.8.4.4
2) Set a fixed IP-address for the Virtual machine
3) added -additionalParameters @("--network=tlan","--ip 172.21.31.10") <- 10 an ip-address for each container
4) Right click in Hyper-V, Settings on Network, Advanced, Enable Mac-address spoofing.

If it still does not work. Then try to find me on tomorrow or friday.

**Comment by Bleika** [comment]
Thanks Palle,

I've finally been able to "solve" this: it seems that the HNS on my localized version of my VM in combination with Docker caused the error...

**I now have an identical VM (exact same build and config) BUT in English.**

Installing Docker the exact same way did not cause an error with HNS when I started the service.
Therefore, Docker was able to create the default networks.

I pulled the same images as before and repeated the last used syntax for navcontainerhelper:
The container was created correctly and connected automatically to the default network.

In the meanwhile I received different suggestions like to use Docker 17.06, but since I have a workaround on my EN VM I'm not trying them out for now - I've already lost a weeks work on this issue and derived problems
